Perf: Replace rtl_tcp linked list with ring buffer

blinick opened the pull request:
Removed the linked list structure, locking, and malloc/free on each USB callback.
Replaced it with a non blocking ring buffer for performance.  This resulted
in a large performance improvement when using an Raspberry Pi or Raspberry Pi Zero W
as a rtl_tcp server.  The sample rate could be doubled and instead of wired ethernet
wifi could be used.  Frequnecy change lag was reduced greatly also.

Signed-off-by: Stephen Blinick <stephen@stesoft.com>

blinick commented:
Blog post with more detail on the code change and results: http://www2.stesoft.com/blog/index.php/2019/02/10/software-defined-radio-with-raspberry-pi-part2/

Thanks!

Hosam-M commented:
hi Stephen thanks for sharing with us i have this problem too but after reading great article Software Defined Radio with Raspberry Pi (Part1) and (Part2) i didnt know how to edit it or make it work can you write the steps or the way how to do it ?

jh-cg commented:
I'm afraid implementing this lock-free ring buffer without explicitly marking atomic access and volatility of head and tail may not work. It is not guaranteed that the variables are written and read in the order you see in the code. See https://lwn.net/Articles/793253/

blinick commented:
@Hosam-M I apologize for not responding to your comment for several months, somehow I didn't get the notification email.     To build off my fork of rtl_sdr, which I think is the only way to get the code at this time, maybe try something like:
```

  git clone https://github.com/blinick/rtl-sdr.git
  cd rtl-sdr/
  git checkout remotes/origin/wip_rtltcp_ringbuf
  mkdir build
  cd build
  cmake ../
  make
  src/rtl_tcp -a 0.0.0.0
```

After which you can install rtl_sdr from this build.    Keep in mind you may need to purge the upstream packages/install first if you've already installed it from somewhere else.

blinick commented:
@jh-cg Thank you for the response!  For what it’s worth, I’ve been able to run this continuously these last several months without any errors or corruption in the byte stream.

In any case, I felt like I had addressed your concerns in the submission by using the correct declaration class specifiers for the variables in question.  It’s even addressed in your linked document, under quiz question #5.     From what I can tell, any reference or manipulation of the ring buffer head/tail will not be optimized away or delayed.     Do you have a specific scenario I can dig into? I’m happy to rewrite as necessary.  This design pattern is quite common in the embedded world where locks are very heavy weight for the CPU due to unsophisticated caches and memory controllers behavior with atomic operations.  In that case, we'd much prefer to have the compiler generate code that may execute more reads/writes than are necessary.

dnesting commented:
Hi @blinick.  When you say you've run this continuously for several months without any corruption, how did you measure this?  What was the CPU architecture?

> I felt like I had addressed your concerns in the submission by using the correct declaration class specifiers

I assume you're referring to your use of `volatile`.  This keyword ensures that when loads and stores occur, they're done directly to/from memory.  It does not ensure *consistency* during execution.  It's intended more for situations like mmap'd files, where the contents of memory will change for reasons completely external to the program and where it's unsafe to cache loads or delay stores.  When I say consistency, I'm referring to things like this:

```bufferleft = ringbuf_sz - ((ringbuf_head < ringbuf_tail) ? (ringbuf_head - ringbuf_tail + ringbuf_sz) : (ringbuf_head - ringbuf_tail));```

Every one of these loads and stores of `ringbuf_head` and `ringbuf_tail` could race against loads and stores in another thread.  Each variable could, at any moment in time, be *inconsistent* with the other variables, and even with *themselves* at different points of execution of this expression.  This code assumes that `ringbuf_head < ringbuf_tail` won't have some instructions from another thread occurring in between the load on the left and the load on the right.  It goes on to assume that the loads in `ringbuf_head - ringbuf_tail` will be consistent with the loads earlier in the expression. 

The way to prevent consistency problems like these is through the use of a consistency-enforcement construct like a mutex.  You need to prevent the other thread from modifying `ringbuf_head` and `ringbuf_tail` while you're trying to do math on them.  Otherwise your math will be all wrong.  You need this sequence of instructions to be done *exclusive* of instructions happening in another thread that change these values.  The `volatile` keyword doesn't do this (and if you're using mutexes, isn't actually buying you anything).

> It’s even addressed in your linked document, under quiz question #5.

This is about compilers being forbidden from *introducing* data races.  But you already have a data race.

blinick commented:
@dnesting  thanks very much for the detailed reply.   

Yes, the architecture I'm referring to is the raspberry pi, I have a couple SDR setups running for a few months now continuously.   The motivation behind this PR is actually specifically for raspberry pi performance, as the it wasn't viable before these changes.  I've had quite a lot of interest in these optimizations after posting about them in the blog I reference above.  It's a pretty compelling use-case, to place the rtl_tcp server in a small SBC attached to the antenna, and run applications elsewhere on a desktop, across wifi.

I haven't had a chance to test/try this code on a IA platform as of yet.  I'd be glad to test on another platform.  Is there a regression for PR's I should be running prior to submission?    

I hope you can understand my confusion prior to your message here, as you're asserting a different concern.  The first comments above by @jh-cg  specifically link to an article about compiler introduced data races, and in your comments you are concluding the issue is more fundamental in how the code is executed, even assuming it's faithfully complied into machine code without optimizations causing further issues.

That said, due to pesky work obligations, I need to consider what you've written and I'll try to provide a response in the near future.  Thanks again for the reply.

dnesting commented:
Here is a simple demonstration if you prefer to see the problem in action:

```c
#include <stdio.h>
#include <pthread.h>

volatile int a, b;  // see note

void* worker() {
	while (1) {
		a++; b++;
	}
}

int main() {
	pthread_t tid;
	pthread_create(&tid, NULL, worker, NULL);

	int i;
	while (1) {
		if (a - b != 0) {
			printf("%d: a==%d b==%d\n", i, a, b);
			break;
		}
		i += 1;
		if (i % 1000000 == 0) {
			printf("%d: nothing seen yet\n", i);
		}
	}
}
```

(Note that `volatile` here doesn't actually change the object code at all; each load/store is done to memory.)

```
$ cc demo.c -o demo
$ ./demo
10248: a==46 b==45
```

> specifically link to an article about compiler introduced data races

The compiler isn't introducing the data race (and as you pointed out earlier, it's *forbidden* from doing so).  It's just exacerbating its effects.  The data race was always there due to assumptions made about the language's memory model, lurking and waiting. You're right though: the article is about some of the more esoteric data races that aren't intuitive even to people who know how to anticipate data races.  But the problem is more fundamental than that.

rod2k commented:
ok, i have tested the code over the same scenario, RTL_TCP over Rpi and using SDRTouch for Android as client, work flawlessly, i know there an issue about data race, but in practical scenario, it doesn't matter, good improve

blinick commented:
I apologize for not getting back to the discussion sooner due to work obligations.

@dnesting thank you for supplying an example that describes your concern, it helps clear things up for me with regards to the specific issues.    Now that I get exactly what you're concerned with I can definitively say the concerns are not an issue with regards to this PR.

Your example clearly describes a program with a data race, where one thread viewing a set of variables asynchronously observes a state for the given variables which appear to be modified in lock-step each loop.       

The misunderstanding here is that your example implicitly defines correctness as `a==b` at all times.  The assumption is that the programmer believes that the evaluation of these variables only occurs discretely at the end of the while loop.   Obviously that's not the case, because variables A & B aren't updated together atomically.  As you stated, a mutex shared between the two threads would ensure that A & B are never observed while being updated. 

The lockless ringbuffer in the PR _assumes_ this data race exists and functions fine with it.  It's totally fine if tail changes asynchronously while the other thread is observing it.  It'll either find data to process (head != tail) or it won't.  If it doesn't, it'll try again next loop.    The types I used are consistent between the threads, and for an ILP64 or ILP32 system load/stores from the register file to memory will be atomic (i.e. no data-tearing issues).  

The PR is all about performance.  In the embedded world -- raspberry pi, or as @rod2k  mentions, android -- mutexes tend to generate lock instructions for the specific architecture.   Lock instructions may be an atomic swap or load/store conditional pair.   In the "non IA" world of things, chip architects don't expend as many gates on elaborate cache architectures or memory controllers.   Lock instructions can therefore be quite costly in terms of pipeline stalls and even impacts to other cores as signals have to punch through multiple layers of cache to prevent data races in the CPU.   It's a common design pattern to implement things this way for performance.  The alternative is to write code that doesn't depend on the mutex.  

Obviously this caused some confusion here, so I can try a rewrite with better comments.  I could perhaps write it more elegantly with unsigned arithmetic so that the expressions being evaluated are shorter also.  I believe what's here is correct but glad to make it more readable for future contributors.     

I'll work on an update for this as soon as I have time.  Thanks again for your review.

mikerobertsn commented:
Can we have this committed, so i can use it in low powered sbc?

steve-m commented:
I don't think completely replacing the linked list with a ringbuffer without still offering the linked list as an option is a good idea. rtl_tcp works fine for years, and the linked list clearly has its benefits, as no data is lost even in high latency networks. However, if the CPU load can really be improved with a ringbuffer, why not offer it as a commandline switch?

mrobertsn commented:
Looking through the testing the OP did, it just seems to have fixed problems which I have encountered too when using different RPI's and low powered Linux pc's over wifi. I've noticed rtl_tcp is always quite flaky, it just seems like any little blip seems to cause stuttering or the dreaded "ll+ now (##)" messages etc. It seems like the modification is making better use of the WiFi link between the devices as I have always found I need a wifi link multiple times faster than the data rate to have error-free operation, but from the OP's testing it seems like he has made progress on this to stream more data across without problems.

I agree, a command-line switch sounds like the best option.

![image](https://user-images.githubusercontent.com/50306399/66723288-f3713880-ee62-11e9-9c79-ce4e53acd269.png)

http://www2.stesoft.com/blog/index.php/2019/02/10/software-defined-radio-with-raspberry-pi-part2/

digi2017 commented:
I have question. Does this patch work also with the rsp_tcp application for the SDRPlay devices also?

val123456 commented:
Thanks for the patch, I tried this on my RPi 4 running rtlamr (both on same RPi) and get this "error" instead of the ll+:

overrun: head=0 tail=262144, Trimming 524288 bytes from tail of buffer
overrun: head=524288 tail=786432, Trimming 524288 bytes from tail of buffer
overrun: head=1048576 tail=1310720, Trimming 524288 bytes from tail of buffer
overrun: head=5505024 tail=5767168, Trimming 524288 bytes from tail of buffer

Stats:
>> [ 4.66MB/s ]  [ bytes_in_flight(cur/max) = 5376K / 5888K ]

Used the checkout/build instructions above.

Jalict commented:
> 
> 
> Thanks for the patch, I tried this on my RPi 4 running rtlamr (both on same RPi) and get this "error" instead of the ll+:
> 
> overrun: head=0 tail=262144, Trimming 524288 bytes from tail of buffer
> overrun: head=524288 tail=786432, Trimming 524288 bytes from tail of buffer
> overrun: head=1048576 tail=1310720, Trimming 524288 bytes from tail of buffer
> overrun: head=5505024 tail=5767168, Trimming 524288 bytes from tail of buffer
> 
> Stats:
> 
> > > [ 4.66MB/s ]  [ bytes_in_flight(cur/max) = 5376K / 5888K ]
> 
> Used the checkout/build instructions above.

I have same same issue
```
Allocating 15 zero-copy buffers
set sample rate 2048000
Allocate 8388608 bytes for ringbuf.
>> [ 0.00MB/s ]  [ bytes_in_flight(cur/max) =  256K /  256K ]
set freq correction 0
set freq 100000000
set agc mode 0
set gain mode 1
set tuner gain by index 0
overrun: head=3932160 tail=4194304, Trimming 524288 bytes from tail of buffer
overrun: head=4456448 tail=4718592, Trimming 524288 bytes from tail of buffer
overrun: head=4980736 tail=5242880, Trimming 524288 bytes from tail of buffer
overrun: head=5505024 tail=5767168, Trimming 524288 bytes from tail of buffer
overrun: head=6029312 tail=6291456, Trimming 524288 bytes from tail of buffer
overrun: head=6553600 tail=6815744, Trimming 524288 bytes from tail of buffer
overrun: head=2883584 tail=3145728, Trimming 524288 bytes from tail of buffer
overrun: head=3407872 tail=3670016, Trimming 524288 bytes from tail of buffer
overrun: head=3932160 tail=4194304, Trimming 524288 bytes from tail of buffer
overrun: head=8126464 tail=0, Trimming 524288 bytes from tail of buffer
overrun: head=262144 tail=524288, Trimming 524288 bytes from tail of buffer
overrun: head=786432 tail=1048576, Trimming 524288 bytes from tail of buffer
```

blinick commented:
This isn’t an error, but more of a status that the network isn’t keeping up with the data being sent.  Usually this is because of network congestion, or a drop-out in WIFI if you’re using that.  Data in the ring buffer between head & tail represents data not yet sent back to the SDR client software.   Since the stream is real-time, the software chooses to overwrite data in the ring buffer that wasn’t transmitted yet.     

It won’t hurt anything, but it means if you’re listening to a channel you’re hear a “skip”.  Prior to the patch, the buffer would simply reallocate more and more memory to hold data in flight.  At that point, you’d end up listening to the stream 10, 20, 30 seconds behind real-time and channel changes would accordingly be late.  Also the dynamic linked-list design added a lot of overhead which slowed down transmission on RPI platforms. 

Since your average below is around 37Mbps, it seems like the RPI4 could push more data and may be having wifi dropouts.  I’d try and improve WIFI signal for the RPI.

> On Apr 9, 2020, at 12:21 AM, Frans Larsen <notifications@github.com> wrote:
> 
> 
> Thanks for the patch, I tried this on my RPi 4 running rtlamr (both on same RPi) and get this "error" instead of the ll+:
> 
> overrun: head=0 tail=262144, Trimming 524288 bytes from tail of buffer
> overrun: head=524288 tail=786432, Trimming 524288 bytes from tail of buffer
> overrun: head=1048576 tail=1310720, Trimming 524288 bytes from tail of buffer
> overrun: head=5505024 tail=5767168, Trimming 524288 bytes from tail of buffer
> 
> Stats:
> 
> [ 4.66MB/s ] [ bytes_in_flight(cur/max) = 5376K / 5888K ]
> 
> Used the checkout/build instructions above.
> 
> I have same same issue
> 
> Allocating 15 zero-copy buffers
> set sample rate 2048000
> Allocate 8388608 bytes for ringbuf.
> >> [ 0.00MB/s ]  [ bytes_in_flight(cur/max) =  256K /  256K ]
> set freq correction 0
> set freq 100000000
> set agc mode 0
> set gain mode 1
> set tuner gain by index 0
> overrun: head=3932160 tail=4194304, Trimming 524288 bytes from tail of buffer
> overrun: head=4456448 tail=4718592, Trimming 524288 bytes from tail of buffer
> overrun: head=4980736 tail=5242880, Trimming 524288 bytes from tail of buffer
> overrun: head=5505024 tail=5767168, Trimming 524288 bytes from tail of buffer
> overrun: head=6029312 tail=6291456, Trimming 524288 bytes from tail of buffer
> overrun: head=6553600 tail=6815744, Trimming 524288 bytes from tail of buffer
> overrun: head=2883584 tail=3145728, Trimming 524288 bytes from tail of buffer
> overrun: head=3407872 tail=3670016, Trimming 524288 bytes from tail of buffer
> overrun: head=3932160 tail=4194304, Trimming 524288 bytes from tail of buffer
> overrun: head=8126464 tail=0, Trimming 524288 bytes from tail of buffer
> overrun: head=262144 tail=524288, Trimming 524288 bytes from tail of buffer
> overrun: head=786432 tail=1048576, Trimming 524288 bytes from tail of buffer
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/osmocom/rtl-sdr/pull/6#issuecomment-611373500>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACXK6OY7CF2QO2M6HYOWAZ3RLVZQZANCNFSM4GWMXN6A>.
>

p8768 commented:
hackrf one

porkchoptree33 commented:
> 
> 
> @Hosam-M I apologize for not responding to your comment for several months, somehow I didn't get the notification email. To build off my fork of rtl_sdr, which I think is the only way to get the code at this time, maybe try something like:
> 
> ```
> 
>   git clone https://github.com/blinick/rtl-sdr.git
>   cd rtl-sdr/
>   git checkout remotes/origin/wip_rtltcp_ringbuf
>   mkdir build
>   cd build
>   cmake ../
>   make
>   src/rtl_tcp -a 0.0.0.0
> ```
> 
> After which you can install rtl_sdr from this build. Keep in mind you may need to purge the upstream packages/install first if you've already installed it from somewhere else.

I'm not certain, but this doesn't seem to work for me when using my RPi 3B (Not 3B+) with a NooElec rtl-SDR. To put it simply, running rtl-TCP with these instructions has only caused me more problems.

1. Whether this is a problem with this fork or the original rtl-TCP, When I connect to the RPi, the spectrum is completely empty unless I reset to 0MHz and re-tune to the frequency I wish to listen to.

2. I'm not sure how you're managing to get 3.2MSPS on the RPi 3B's Built-in Wi-Fi, My home router is more than capable of handling over 800Mbits of data from a single client, however I'm having the same byte data overrun warnings mention previously by other users (Ex: **overrun: head=XXXXXXX tail=XXXXXXX, Trimming XXXXXX bytes from tail of buffer**) that you said indicated the network couldn't handle the data (I even moved my RPi right next to my router and I was still having overrun warnings and 4+ second lag on 1.024MSPS), Any ideas what may be causing it?

3. There is very clearly some random spikes of RF interference following the visible spectrum wherever it goes, Naturally I would think this is actual interference from in-home devices such as power supplies, however, when using it locally on my pc in the same location and with the original rtl-TCP (Not this fork), I don't have the RF issues, any ideas what could be causing this? (The interference spikes are VERY persistent and never move, it's just multiple tiny spikes of RF interference.)

porkiedev commented:
(Apologies if you got a notification of the exact same reply from another account, I accidentally replied on my old GitHub account which I no longer use and I want to receive notifications on my personal email instead of my old email.)

> 
> 
> @Hosam-M I apologize for not responding to your comment for several months, somehow I didn't get the notification email. To build off my fork of rtl_sdr, which I think is the only way to get the code at this time, maybe try something like:
> 
> ```
> 
>   git clone https://github.com/blinick/rtl-sdr.git
>   cd rtl-sdr/
>   git checkout remotes/origin/wip_rtltcp_ringbuf
>   mkdir build
>   cd build
>   cmake ../
>   make
>   src/rtl_tcp -a 0.0.0.0
> ```
> 
> After which you can install rtl_sdr from this build. Keep in mind you may need to purge the upstream packages/install first if you've already installed it from somewhere else.

I'm not certain, but this doesn't seem to work for me when using my RPi 3B (Not 3B+) with a NooElec rtl-SDR. To put it simply, running rtl-TCP with these instructions has only caused me more problems.

1. Whether this is a problem with this fork or the original rtl-TCP, When I connect to the RPi, the spectrum is completely empty unless I reset to 0MHz and re-tune to the frequency I wish to listen to.

2. I'm not sure how you're managing to get 3.2MSPS on the RPi 3B's Built-in Wi-Fi, My home router is more than capable of handling over 800Mbits of data from a single client, however I'm having the same byte data overrun warnings mention previously by other users (Ex: overrun: head=XXXXXXX tail=XXXXXXX, Trimming XXXXXX bytes from tail of buffer) that you said indicated the network couldn't handle the data (I even moved my RPi right next to my router and I was still having overrun warnings and 4+ second lag on 1.024MSPS), Any ideas what may be causing it?

3. There is very clearly some random spikes of RF interference following the visible spectrum wherever it goes, Naturally I would think this is actual interference from in-home devices such as power supplies, however, when using it locally on my pc in the same location and with the original rtl-TCP (Not this fork), I don't have the RF issues, any ideas what could be causing this? (The interference spikes are VERY persistent and never move, it's just multiple tiny spikes of RF interference.)

bogdankoval commented:
> After which you can install rtl_sdr from this build. Keep in mind you may need to purge the upstream packages/install first if you've already installed it from somewhere else.

Greate news! I hope it still to work with direct sampling mod of rtl_tcp. I will check it soon.

Jasoroony commented:
I think the race condition (mentioned much earlier) would be solved by changing the logic to drop latest data that won't fit into the ring buffer.  This way, one thread only ever updates the head, the other thread only ever updates the tail.  At the moment, one of the threads can update both, to allow dropping the oldest data - which I think is where the concern was.

uriva9w commented:
> Stephen

Dear Stephen. I have a lot of ll+ when I run rtlamr -msgtype=r900 as you described it in your blog. Could you please post a step-by step procedure to improve the rtl_tcp. Thx
